### ISSUE TITLE
Unify REPL and debugging experience

### ISSUE THREAD

**Issue by int19h** [Microsoft/PTVS#751]
There is a lot of commonality between a REPL and a debugged application. In particular, both are a front-end to some running process with live state that can be inspected and modified. Debugger has UI-centric experience for it today in the form of Locals and Watch windows, as well as Debug Interactive. On the other hand, REPL doesn't have any equivalent for Locals (but see #81).

Ideally, these two experiences should be identical, or at least as close to each other as possible. Whether it is a process that is paused at a breakpoint, or paused waiting for REPL input, there should be convenient UI to inspect and edit variables in the current context, and all the advanced visualizer functionality (#747, #748) should tie into it in the same manner. If that is impossible to do with the existing debugger windows due to the lack of extensibility points, then any new facilities that are provided in their stead should similarly behave identically in both scenarios.

Furthermore, if a breakpoint is set in a file that is then evaluated somehow from the REPL (e.g. by importing it), that breakpoint should be hit immediately, without the need to do `$attach` etc - in other words, the REPL should always be treated as an active debugger session (similar to e.g. R).

**Comment by zooba** [comment]
This probably ties in with the ptvsd refactor #750 as well, if we can design a way to switch debugging on and off easily and avoid huge performance loss in the 'off' state. The nice thing about the REPL is that we don't really need to keep a lot of state between each input.

Then we could automatically turn debugging on when any breakpoints are set (maybe display some UI flag as well to show that we're debugging). Ideally, I think we'd want to let people use #81, #747 and #748 in place of the regular debugging UI anyway, and I'm sure everyone will want to.

**Comment by zooba** [comment]
From #220:

The following suggestions would substantially enhance the development speed, especially for scientific computing.

1. When selecting code in the editor and right clicking: "Send to Debug Interactive" option is missing. only "Send to Interactive" exists. Having such an option (+ keyboard shortcut) would allow throwing chunks of code into evaluation much faster.

The next two bullets I suggest are useful for the common scenario where one wishes to put aside in memory some variables that take a while to prepare (long calculation, or loading from large file) and use them for debugging a new function that is being written and receives these variables as input.
 One could then prepare the variables once, place a breakpoint in the half written function, inspect local variables and continue adding code. The process repeats until the function is completed.
 This is supported in Matlab for example.

1. Allowing global variables that don't disappear between debugging sessions.

2. Calling a function from REPL should stop on breakpoints within that function. (currently, the breakpoints are skipped)

**Comment by nidleo** [comment]
I strongly agree with what has been said above - joining the release and debug experiences, especially with the part about no performance loss in "debug off" state. I believe this is crucial as it's very natural to use the REPL as your "release mode", which must be as fast as possible.

Until then, I'd like to suggest a couple of small features to enhance current experience:
1. "Send to relevant interactive" command to send selected code to the correct interactive window - debug or the regular one, instead of having to juggle two separate commands.
2. auto-completion of an object's attributes in REPL should be based on the actual object rather than intellisense which may suggest non-existent attributes or miss existing ones.

**Comment by int19h** [comment]
We actually already do Intellisense based on live object inspection in all REPL windows by default (it's configurable in Tools -> Options -> Python Tools).

**Comment by zooba** [comment]
And there's an option to completely disable suggestions that aren't from the live objects. Currently we merge them by default, since we have slightly more information in the editor case (e.g. is this a function or an attribute? When it's live, we can't tell without accessing the value, which may have side-effects).

There are going to be some changes to how debug interactive works regardless of what we decide to do, as VS is working on stuff for their next release. We really want to support the "start debugging interactive code from the interactive" scenario, while VS is going to handle the "I need something better than the immediate window while I'm debugging my project". Basically - we want to improve the standard REPL window to the point where you wouldn't even consider the debug interactive here. Though it is still likely we'll end up with multiple commands.

**Comment by zooba** [comment]
@int19h This is something I'd like you to work on. No specific target release, but let's get it done in pieces that can release immediately - the fewer breaking changes to `ptvsd` the better.

There's also #2006 to cover the suggestions of "make it more obvious which REPL window will get data sent to it", so that doesn't have to be done as part of this.
